torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -2960,23 +2966,73 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::pointToPoint(
 
   auto device = getDevice(tensor);
   std::string key;
-  int p2pRank = 0, p2pTargetRank = 0;
-  bool isSendRecvSelf = false;
+  int p2pRank;
+  int p2pTargetRank;
+  const bool isSendRecvSelf = rank_ == peer;
   // For batch_isend_irecv, ncclGroupStart() would be called upfront
   bool batchP2P = ncclActiveGroupCounter_ > 0;
-  if (batchP2P) {
-    // For batch P2P, we need to treat it like a collective when selecting
-    // communicator, because other ranks can call into this batch other than my
-    // rank and my peer
+  std::shared_ptr<NCCLComm> ncclComm = nullptr;
+  if (this->eagerInit) {
+    /* In eagerInit mode, reuse the parent comm.  Do not lazily create
+     * p2p communicators. */
+    bool showSerializationWarning =
+        getCvarBool(TORCH_NCCL_SHOW_EAGER_INIT_P2P_SERIALIZATION_WARNING, true);
+    if (!batchP2P && showSerializationWarning) {
+      TORCH_WARN_ONCE(
```

Comment:
Telling users to create 2-rank PGs for P2P is bad UX. It seems OK to tell users to create a separate PG for collectives and for P2P, but a P2P PG should be able to send messages to any rank.

I'm concerned that we are dissuading users from adopting eager-init because we couple it with a perf-degradation on P2P without a worsened UX on PG management side.

Reply:
I agree this is not perfect, as we discussed offline. in the existing OSS main branch, eager mode does not even work with p2p because of the 'collective' nature of ncclCommSplit. This PR makes it work, but warn the users about the perf indications.